RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitProducers.cc

```diff
@@ -0,0 +1,137 @@
+    device::ESGetToken<hgcalrechit::HGCalCalibParamDevice, HGCalModuleConfigurationRcd> calibToken_;
+    device::ESGetToken<hgcalrechit::HGCalConfigParamDevice, HGCalModuleConfigurationRcd> configToken_;
+    const device::EDPutToken<hgcalrechit::HGCalRecHitDevice> recHitsToken_;
+    const HGCalRecHitCalibrationAlgorithms calibrator_;
+    int n_hits_scale;
+  };
+
+  HGCalRecHitsProducer::HGCalRecHitsProducer(const edm::ParameterSet& iConfig)
+      : digisToken_{consumes<hgcaldigi::HGCalDigiHost>(iConfig.getParameter<edm::InputTag>("digis"))},
+        calibToken_{esConsumes(iConfig.getParameter<edm::ESInputTag>("calibSource"))},
+        configToken_{esConsumes(iConfig.getParameter<edm::ESInputTag>("configSource"))},
+        recHitsToken_{produces()},
+        calibrator_{iConfig.getParameter<int>("n_blocks"), iConfig.getParameter<int>("n_threads")},
+        n_hits_scale{iConfig.getParameter<int>("n_hits_scale")} {}
+
+  void HGCalRecHitsProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("digis", edm::InputTag("hgcalDigis", "DIGI", "TEST"));
+    desc.add("calibSource", edm::ESInputTag{})->setComment("Label for calibration parameters");
+    desc.add("configSource", edm::ESInputTag{})->setComment("Label for ROC configuration parameters");
+    desc.add<int>("n_blocks", -1);
+    desc.add<int>("n_threads", -1);
+    desc.add<int>("n_hits_scale", -1);
+    descriptions.addWithDefaultLabel(desc);
+  }
+
+  void HGCalRecHitsProducer::produce(device::Event& iEvent, device::EventSetup const& iSetup) {
+    auto& queue = iEvent.queue();
+
+    // Read digis
+    auto const& deviceCalibParamProvider = iSetup.getData(calibToken_);
+    auto const& deviceConfigParamProvider = iSetup.getData(configToken_);
+    auto const& hostDigisIn = iEvent.get(digisToken_);
+
+    //printout new conditions if available
+    LogDebug("HGCalCalibrationParameter").log([&](auto& log) {
+      if (calibWatcher_.check(iSetup)) {
+        for (int i = 0; i < deviceConfigParamProvider.view().metadata().size(); i++) {
+          log << "idx = " << i << ", "
+              << "gain = " << deviceConfigParamProvider.view()[i].gain() << ","
+              << "ADC_ped = " << deviceCalibParamProvider.view()[i].ADC_ped() << ", "
+              << "CM_slope = " << deviceCalibParamProvider.view()[i].CM_slope() << ", "
+              << "CM_ped = " << deviceCalibParamProvider.view()[i].CM_ped() << ", "
+              << "BXm1_slope = " << deviceCalibParamProvider.view()[i].BXm1_slope() << ", " << std::endl;
+        }
+      }
+    });
+
+    int oldSize = hostDigisIn.view().metadata().size();
+    int newSize = oldSize * n_hits_scale;
+    auto hostDigis = HGCalDigiHost(newSize, queue);
+    // TODO: replace with memcp ?
+    for (int i = 0; i < newSize; i++) {
+      hostDigis.view()[i].tctp() = hostDigisIn.view()[i % oldSize].tctp();
```

Comment:
I would suggest to rewrite this along the lines of
```suggestion
    for (int i = 0; i < oldSize; i++) {
      for (int j = 0; i < n_hits_scale; i++) {
        int k = j * oldSize + i;
        hostDigis.view()[k].tctp() = hostDigisIn.view()[i].tctp();
        ...
```
to improve the locality of the reads.

But I don't know if it matters in terms of actual performance.